cpp/src/traversal/opg/frontier_expansion.cu

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/handle.hpp>
+#include "load_balance.cuh"
+#include "bfs_comms.cuh"
+
+namespace detail {
+
+namespace opg {
+
+struct bitwise_or {
+  __device__ unsigned operator()(unsigned& a, unsigned & b) { return a | b; }
+};
+
+struct remove_visited {
+  __device__ unsigned operator()(unsigned& visited, unsigned& output) {
+    //OUTPUT AND VISITED - common bits between output and visited
+    //OUTPUT AND (NOT (OUTPUT AND VISITED))
+    // - remove common bits between output and visited from output
+    return (output & (~( output & visited )));
+  }
+};
+
+template <typename VT>
+struct bitwise_atomic_or {
+  unsigned * output_frontier_;
+  VT * predecessors_;
+
+  bitwise_atomic_or(
+      unsigned * output_frontier,
+      VT * predecessors) :
+    output_frontier_(output_frontier),
+    predecessors_(predecessors) {}
+
+  __device__ void operator()(VT src, VT dst) {
+    unsigned active_bit = static_cast<unsigned>(1)<<(dst % BitsPWrd<unsigned>);
+    unsigned prev_word =
+      atomicOr(output_frontier_ + (dst/BitsPWrd<unsigned>), active_bit);
+    //If this thread activates the frontier bitmap for a destination
+    //then the source is the predecessor of that destination
+    if (prev_word & active_bit == 0) {
+      predecessors_[dst] = src;
+    }
+  }
+};
+
+struct is_not_equal {
+  unsigned cmp_;
+  unsigned * flag_;
+  is_not_equal(unsigned cmp, unsigned * flag) : cmp_(cmp), flag_(flag) {}
+  __device__ void operator()(unsigned& val) {
+    if (val != cmp_) {
+      *flag_ = 1;
+    }
+  }
+};
+
+template <typename VT, typename ET, typename WT>
+void bfs(raft::handle_t const &handle,
+    cugraph::experimental::GraphCSRView<VT, ET, WT>& graph,
+    VT *predecessors,
+    const VT start_vertex) {
```

Comment:
How does this partition work/data across GPUs? What are the expectations on the input graph?

Reply:
For now the graph is 1d partitioned